### PR TITLE
brancher: make brancher part of scm

### PR DESCRIPTION
**Pull request by pared**
* [x] ❗ I have followed the [Contributing to DVC](https://dvc.org/doc/user-guide/contributing/core) checklist.

* [x] 📖 If this PR requires [documentation](https://dvc.org/doc) updates, I have created a separate PR (or issue, at least) in [dvc.org](https://github.com/iterative/dvc.org) and linked it here.

Thank you for the contribution - we'll try to review it as soon as possible. 🙏

Prerequisite for #4446 

In order to make plots/metrics/params `diff` accept any viable target, we need to make brancher repo agnostic. Considering that `diff` makes sense only in case of some kind of `scm`, it makes sense to make brancher part of `scm`.

### CODE REVIEW

**Comment by pared** [comment]
@Suor does the viewer interact with brancher in any way? In this change I intended to unify the branches behaviour and return "workspace" instead of ""/"workspace" depending on repository status, not sure if that won't interfere with something at `viewer`.

**Review by pmrowla — APPROVED** [review]

File: tests/func/metrics/test_show.py [hunk 1]
```diff
@@ -17,23 +17,25 @@ def test_show_simple(tmp_dir, dvc, run_copy_metrics):
     run_copy_metrics(
         "metrics_t.yaml", "metrics.yaml", metrics=["metrics.yaml"],
     )
-    assert dvc.metrics.show() == {"": {"metrics.yaml": 1.1}}
+    assert dvc.metrics.show() == {"workspace": {"metrics.yaml": 1.1}}
```
Comment (efiop) [inline tests/func/metrics/test_show.py:20]: This is a breaking API change, people might rely on this, so normally we can't just break it like that, unless we have a very good reason.

**Review by efiop — CHANGES_REQUESTED** [review]
This is breaking our API, could you clarify the reasons to do that?

File: dvc/scm/base.py [hunk 2]
```diff
@@ -165,3 +166,15 @@ def belongs_to_scm(self, path):
...
+
+    def brancher(
+        self,
+        root_dir,
+        revs=None,
+        all_branches=False,
+        all_tags=False,
+        all_commits=False,
+    ):  # pylint: disable=no-self-use, unused-argument
+        yield "workspace", LocalTree(
```
Comment (Suor) [inline dvc/scm/base.py:178]: Do we need to return workspace from here? It looks like it makes mess, i.e. we need to skip it later.

**Comment by Suor** [comment]
Another thing is a branch or tag might be named workspace. In this case on metric show that branch is indistinguishable from actual workspace.

File: dvc/scm/base.py (hunk 2 above)
Comment (efiop) [reply dvc/scm/base.py:178]: Also `LocalTree` has absolutely nothing to do with Base SCM

**Comment by pared** [comment]
Seems like starting from redesigning the brancher was wrong idea. Not only it would break the current functionality, but also the end result could be achieved in [other way](https://github.com/iterative/dvc/pull/4627). Closing for now.

**Pull request closed without merging**